yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,785 @@
+import { type IndexedTreeId, MerkleTreeId, type MerkleTreeReadOperations, getTreeHeight } from '@aztec/circuit-types';
+import { NullifierLeafPreimage, PublicDataTreeLeafPreimage } from '@aztec/circuits.js';
+import { poseidon2Hash } from '@aztec/foundation/crypto';
+import { Fr } from '@aztec/foundation/fields';
+import { type IndexedTreeLeafPreimage, type TreeLeafPreimage } from '@aztec/foundation/trees';
+
+import cloneDeep from 'lodash.clonedeep';
+
+/****************************************************/
+/****** Structs Used by the AvmEphemeralForest ******/
+/****************************************************/
+
+/**
+ * The preimage and the leaf index of the Low Leaf (Low Nullifier or Low Public Data Leaf)
+ */
+type PreimageWitness<T extends IndexedTreeLeafPreimage> = {
+  preimage: T;
+  index: bigint;
+  update: boolean;
+};
```

Comment:
I still think this name isn't great. Elsewhere we use "witness" in the context of merkle trees to be leaf index & sibling path. I don't know what a better name is, but PreimageWitness doesn't feel right